### PR TITLE
pkg/openwsn: add ieee802154_hal based radio

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds the option to use OpenWSN on top of the `ieee802154_hal` which enables using cc2538 and nrf52840 radios.

### Testing procedure

Run `tests/pkg_openwsn` on a cc2538 and a nr52840 based boards.

- cc2538_rf

```
17:53:22 SUCCESS 35d5 [OPENWSN] booted
17:53:27 SUCCESS Setting mote 3248 as root
17:53:27 ERROR 3248 [IEEE802154E] wrong state 1 in startSlot, at slotOffset 1
17:53:28 INFO registering DAGroot 96-68-31-23-96-16-23-37
17:55:02 SUCCESS 35d5 [OPENWSN] booted
17:55:04 WARNING MoteProbe@/dev/riot/tty-openmote-b_1: invalid serial frame: (10B) 7e-53-35-d5-01-00-fe-ca-35-7e wrong CRC
17:55:04 SUCCESS 891 [OPENWSN] booted
17:55:41 SUCCESS 891 [IEEE802154E] synchronized at slotOffset 0
17:55:49 INFO 891 [SIXTOP] sending a 6top request
17:55:51 WARNING 891 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 1)
17:55:52 WARNING 891 [IEEE802154E] large timeCorr.: 6 ticks (code loc. 0)
17:55:52 SUCCESS 891 [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
17:55:58 INFO received RPL DAO from bbbb:0:0:0:3696:31f6:2396:6823
        - parents:
           bbbb:0:0:0:9668:3123:9616:2337
        - children:
17:55:58 WARNING 891 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 1)
17:55:59 WARNING 891 [IEEE802154E] large timeCorr.: 6 ticks (code loc. 0)
```

- nrf52840-mdk
 
```

22:24:39 SUCCESS 161c [OPENWSN] booted
22:24:39 WARNING unregistering DAGroot d6-bf-6c-f1-fc-da-3c-e4
22:24:50 SUCCESS Setting mote 161c as root
22:24:50 ERROR 161c [IEEE802154E] wrong state 1 in startSlot, at slotOffset 1
22:24:50 INFO registering DAGroot d6-bf-6c-f1-fc-da-3c-e4
22:29:07 INFO received RPL DAO from bbbb:0:0:0:daf2:7e75:9bf8:6a9b
        - parents:
           bbbb:0:0:0:d6bf:6cf1:fcda:3ce4
        - children:
22:29:25 INFO received RPL DAO from bbbb:0:0:0:daf2:7e75:9bf8:6a9b
        - parents: 
           bbbb:0:0:0:d6bf:6cf1:fcda:3ce4
        - children:
22:29:31 INFO received RPL DAO from bbbb:0:0:0:daf2:7e75:9bf8:6a9b
        - parents:
           bbbb:0:0:0:d6bf:6cf1:fcda:3ce4
        - children:
22:30:52 INFO received RPL DAO from bbbb:0:0:0:daf2:7e75:9bf8:6a9b
        - parents:
           bbbb:0:0:0:d6bf:6cf1:fcda:3ce4
        - children:
22:32:07 INFO received RPL DAO from bbbb:0:0:0:daf2:7e75:9bf8:6a9b
        - parents:
           bbbb:0:0:0:d6bf:6cf1:fcda:3ce4

```

### Issues/PRs references

Depends on #15144 and #15140

### CODE REVIEW

**Comment by fjmolinas** [comment]
NOTE: still need to measure for this board

```
#define PORT_maxTxDataPrepare   (3355 / PORT_US_PER_TICK )  /* ~110 ticks at @32768Hz */
#define PORT_maxRxAckPrepare    (610 / PORT_US_PER_TICK )   /* ~20 ticks at @32768Hz */
#define PORT_maxRxDataPrepare   (1000 / PORT_US_PER_TICK )  /* ~33 ticks at@32768Hz */
#define PORT_maxTxAckPrepare    (1525 / PORT_US_PER_TICK )  /* ~50 ticks at@32768Hz */
#define PORT_delayTx            (300 / PORT_US_PER_TICK )   /* ~10 ticks at@32768Hz */
#define PORT_delayRx            (0 / PORT_US_PER_TICK )     /* ~0 ticks at@32768Hz */
```

File: boards/openmote-b/include/board.h [hunk 1]
```diff
@@ -86,6 +86,27 @@
...
+#define OPENWSN_LEDPIN_ERROR    LED0_PIN
+#define OPENWSN_LEDPIN_SYNC     LED1_PIN
+#define OPENWSN_LEDPIN_RADIO    LED3_PIN
+#define OPENWSN_LEDPIN_DEBUG    LED2_PIN
+/** @} */
+
+/**
+ * @name    OpenWSN debugpins configuration
+ * @{
+ */
```
Comment (fjmolinas) [inline boards/openmote-b/include/board.h:102]: Add a comment saying this matches OpenWSN-fw led configuration for this board.

File: pkg/openwsn/Makefile.dep [hunk 2]
```diff
@@ -50,6 +48,33 @@ ifneq (,$(filter openwsn_crypto,$(USEMODULE)))
...
 endif
 
+ifneq (,$(filter openwsn_radio,$(USEMODULE)))
+  ifeq (,$(filter openwsn_radio_%,$(USEMODULE)))
+    ifneq (,$(filter cc2538,$(CPU)))
+      USEMODULE += openwsn_radio_hal
+    endif
+    ifneq (,$(filter nrf5284%,$(CPU_MODEL)))
+      USEMODULE += openwsn_radio_hal
+    endif
```
Comment (jia200x) [inline pkg/openwsn/Makefile.dep:58]:
```suggestion
    ifneq (,$(filter cc2538_rf nrf802154,$(CPU)))
      USEMODULE += openwsn_radio_hal
    endif
```

File: pkg/openwsn/contrib/eui64.c [hunk 3]
```diff
@@ -29,10 +29,22 @@
...
 {
+#ifdef MODULE_OPENWSN_RADIO_HAL
+    if (!_eui64_is_set) {
+        luid_get_eui64(&_eui64);
+        _eui64_is_set = true;
+    }
+    memcpy(addressToWrite, _eui64.uint8, sizeof(_eui64.uint8));
+#else
     eui64_t eui64;
-
```
Comment (jia200x) [inline pkg/openwsn/contrib/eui64.c:35]: unrelated change

File: pkg/openwsn/contrib/openwsn.c [hunk 4]
```diff
@@ -52,46 +61,42 @@ kernel_pid_t openwsn_get_pid(void)
...
-    dev->driver->get(dev, NETOPT_ADDRESS, addr, IEEE802154_SHORT_ADDRESS_LEN);
-    open_addr_t id;
-    id.type = ADDR_16B;
-    memcpy(&id.addr_16b, addr, IEEE802154_SHORT_ADDRESS_LEN);
-    idmanager_setMyID(&id);
-}
+    void* dev = NULL;
+    /* avoid cppcheck style (redundantAssignment)*/
+    (void) dev;
+#ifdef MODULE_OPENWSN_RADIO_NETDEV
```
Comment (jia200x) [inline pkg/openwsn/contrib/openwsn.c:69]: Here and below
```suggestion
#if IS_USED(MODULE_OPENWSN_RADIO_NETDEV)
```

File: pkg/openwsn/contrib/radio_hal.c [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+{
+    (void)dev;
+
+    debugpins_isr_set();
+    _txrx_event_capture_time = sctimer_readCounter();
+    debugpins_isr_clr();
+
+    switch (status) {
+    case IEEE802154_RADIO_CONFIRM_TX_DONE:
+        ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
```
Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:86]: someone should call `ieee802154_radio_confirm_transmit()` on `TX_DONE`. Although most radios do "nothing" there, this is required by the Abstract State Machine of the Radio HAL. Some radios might use it to synchronize internal state variables

Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:86]:
```suggestion
        ieee802154_radio_confirm_transmit(openwsn_radio.dev, NULL);
        ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
```

File: pkg/openwsn/contrib/radio_hal.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+    /* If the radio is still not in TRX_OFF state, spin */
+    while (ieee802154_radio_confirm_on(dev) == -EAGAIN) {}
+
+    /* Enable basic mode, no AUTOACK. no CSMA */
+    ieee802154_radio_set_rx_mode(dev, IEEE802154_RX_AACK_DISABLED);
+    ieee802154_radio_set_csma_params(dev, NULL, -1);
+
+    /* MAC layer will handle retransmissions */
+    if (ieee802154_radio_has_frame_retrans(dev)) {
```
Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:131]: From the Radio HAL API, `set_csma_params`:
```c
     * @param[in] retries number of CSMA-CA retries. If @p retries < 0,
     *                    retransmissions with CSMA-CA MUST be disabled.
     *                    If @p retries == 0, the @ref
     *                    ieee802154_radio_ops::request_transmit function is
     *                    equivalent to CCA send.

```

This means `set_csma_params` will already disable retransmissions. The reason is because hardware retransmission software don't support "retransmissions without CSMA-CA". Therefore this lines can be removed

File: pkg/openwsn/contrib/radio_hal.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+        ieee802154_radio_set_frame_retrans(dev, 0);
+    }
+#ifdef MODULE_CC2538_RF
+    /* If frame filtering is enabled cc2538 will not accept beacons
+       where the destination-address mode is 0 (no destination address).
+       per rfc8180 4.5.1 the destination address must be set, which means
+       the destination-address mode can't be 0 */
+    ieee802154_radio_set_rx_mode(dev, IEEE802154_RX_PROMISC);
+#endif
```
Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:140]: This can be done with standard if-else

Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:140]:
```suggestion
if (IS_USED(MODULE_CC2538_RF)) {
    /* If frame filtering is enabled cc2538 will not accept beacons
       where the destination-address mode is 0 (no destination address).
       per rfc8180 4.5.1 the destination address must be set, which means
       the destination-address mode can't be 0 */
    ieee802154_radio_set_rx_mode(dev, IEEE802154_RX_PROMISC);
}
```

File: pkg/openwsn/contrib/radio_hal.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+      .page = CONFIG_IEEE802154_DEFAULT_CHANNEL,
+      .pow = CONFIG_IEEE802154_DEFAULT_TXPOWER };
+
+    ieee802154_radio_config_phy(openwsn_radio.dev, &conf);
+}
+
+void radio_rfOn(void)
+{
+    ieee802154_radio_request_on(openwsn_radio.dev);
+    ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
```
Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:188]: Two things here:
1) The `ieee802154_radio_request_set_trx_state` function doesn't have its counterpart `ieee802154_radio_confirm_set_trx_state`. This might produce an inconsistent state in some Radio HAL implementations
2) Anyway, the HAL API guarantees that the is in `TRX_OFF` after `ieee802154_radio_confirm_on`. So, you can remove this line anyway

File: pkg/openwsn/contrib/radio_hal.c [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+            _txrx_event_capture_time = sctimer_readCounter();
+            debugpins_isr_clr();
+            openwsn_radio.startFrame_cb(_txrx_event_capture_time);
+        }
+    }
+}
+
+void radio_rxEnable(void)
+{
+    int ret = ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
```
Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:267]: tip here: Usually going from `TX_ON` to `RX_ON` is way faster than `TRX_OFF` to `RX_ON` (sometimes a couple of microsecs). Maybe you could set the state to TX_ON here

Comment (jia200x) [reply pkg/openwsn/contrib/radio_hal.c:267]: alternatively, I was thinking that adding a `IEEE802154_RX_DISABLED` to `set_rx_mode` could help to disable the detection of SFD while the RX circuit is still on. What do you think?

File: pkg/openwsn/contrib/radio_hal.c [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+                            uint8_t maxBufLen,
+                            int8_t *rssi,
+                            uint8_t *lqi,
+                            bool *crc)
+{
+    ieee802154_rx_info_t rx_info;
+    size_t size = ieee802154_radio_read(openwsn_radio.dev, bufRead,
+                                        maxBufLen, &rx_info);
+
+    /* FCS is skipped by the radio-hal in the returned length, but
```
Comment (jia200x) [inline pkg/openwsn/contrib/radio_hal.c:308]: I guess this is missing in the Radio HAL API documentation

**Comment by jia200x** [comment]
I wrote some comments there, but overall looks good. I will continue with tests

**Comment by fjmolinas** [comment]
@jia200x since you told me you had a couple of `remote-rev%` I wanted to test it, It seems that my board (with missing antenna) has a horrible RSSI and therefore fails to operate correctly (In OpenWSN neighbors or messages from potential neighboers with bad RSSI are discarded), so I would suggest testing with another `BOARD` or adding a proper antenna (I do not have one). I tested with `cc2538dk` and `openmote-b`.

```
12:56:06 SUCCESS 7d12 [IEEE802154E] synchronized at slotOffset 0
12:56:06 SUCCESS 5de6 [IEEE802154E] synchronized at slotOffset 0
12:56:16 INFO 5de6 [CJOIN] sending CJOIN request
12:56:17 VERBOSE New node: ae8d:fee1:60c7:cb8a. Creating new OSCORE context in /home/francisco/workspace/RIOT/bin/oscore_context_ae8dfee160c7cb8a.json.
12:56:17 VERBOSE received JRC join request
12:56:17 SUCCESS 5de6 [CJOIN] node joined
12:56:18 INFO 7d12 [CJOIN] sending CJOIN request
12:56:19 VERBOSE New node: ae8d:fee1:e052:d7d7. Creating new OSCORE context in /home/francisco/workspace/RIOT/bin/oscore_context_ae8dfee1e052d7d7.json.
12:56:19 VERBOSE received JRC join request
12:56:19 SUCCESS 7d12 [CJOIN] node joined
12:56:54 INFO 7d12 [SIXTOP] sending a 6top request
12:56:56 SUCCESS 7d12 [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
12:57:01 INFO 5de6 [SIXTOP] sending a 6top request
12:57:02 SUCCESS 5de6 [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
12:57:06 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
12:57:30 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:e052:d7d7
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
12:57:47 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
12:58:07 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
```

I'll know address all your comments

File: pkg/openwsn/contrib/radio_hal.c (hunk 10 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:308]: resolved

File: pkg/openwsn/contrib/radio_hal.c (hunk 9 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:267]: @jia200x I would prefer the second approach since its explicit, but for now I can go with your first suggestion!

File: pkg/openwsn/contrib/radio_hal.c (hunk 8 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:188]: I think I had it because the `cc2538` does not have a proper on function (and I was lazy to implement it (or better said would rather do it another PR))

Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:188]: I'll add a comment for the momment if you don't mind

File: pkg/openwsn/contrib/openwsn.c (hunk 4 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/openwsn.c:69]: Can't do this, it will complain about `at86rf2xx_dev` not existing since that one is guarded by an `ifdef` (and I don't think I can do it differently without compiler complaints)

File: pkg/openwsn/contrib/eui64.c (hunk 3 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/eui64.c:35]: removed

File: pkg/openwsn/contrib/radio_hal.c (hunk 5 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:87]: done

File: pkg/openwsn/contrib/radio_hal.c (hunk 9 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:268]: done

**Comment by fjmolinas** [comment]
@jia200x I addressed your changes and restested.

File: pkg/openwsn/contrib/radio_hal.c (hunk 7 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c]: done

Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c]: done

File: pkg/openwsn/contrib/radio_hal.c (hunk 8 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c]: did just that

File: pkg/openwsn/contrib/radio_hal.c (hunk 6 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c]: removed

File: pkg/openwsn/contrib/radio_hal.c (hunk 5 above)
Comment (fjmolinas) [reply pkg/openwsn/contrib/radio_hal.c:87]: done

File: pkg/openwsn/Makefile.dep (hunk 2 above)
Comment (fjmolinas) [reply pkg/openwsn/Makefile.dep]: done

File: boards/openmote-b/include/board.h (hunk 1 above)
Comment (fjmolinas) [reply boards/openmote-b/include/board.h]: done

**Comment by fjmolinas** [comment]
Here are the test results @jia200x if you do not have one of the boards:

<details><summary>3 cc2538 based boards</summary>

```
 ping6 -s 40 -c 5 -i 5 bbbb:0:0:0:ae8d:fee1:e052:d7d7
PING bbbb:0:0:0:ae8d:fee1:e052:d7d7(bbbb::ae8d:fee1:e052:d7d7) 40 data bytes
48 bytes from bbbb::ae8d:fee1:e052:d7d7: icmp_seq=1 ttl=64 time=2171 ms
48 bytes from bbbb::ae8d:fee1:e052:d7d7: icmp_seq=2 ttl=64 time=1204 ms
48 bytes from bbbb::ae8d:fee1:e052:d7d7: icmp_seq=3 ttl=64 time=2259 ms
48 bytes from bbbb::ae8d:fee1:e052:d7d7: icmp_seq=4 ttl=64 time=1294 ms
48 bytes from bbbb::ae8d:fee1:e052:d7d7: icmp_seq=5 ttl=64 time=327 ms

--- bbbb:0:0:0:ae8d:fee1:e052:d7d7 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 20008ms
rtt min/avg/max/mdev = 327.384/1451.077/2258.505/709.834 ms

ping6 -s 40 -c 5 -i 5 bbbb:0:0:0:ae8d:fee1:60c7:cb8a
PING bbbb:0:0:0:ae8d:fee1:60c7:cb8a(bbbb::ae8d:fee1:60c7:cb8a) 40 data bytes
48 bytes from bbbb::ae8d:fee1:60c7:cb8a: icmp_seq=1 ttl=64 time=176 ms
48 bytes from bbbb::ae8d:fee1:60c7:cb8a: icmp_seq=2 ttl=64 time=1217 ms
48 bytes from bbbb::ae8d:fee1:60c7:cb8a: icmp_seq=3 ttl=64 time=264 ms
48 bytes from bbbb::ae8d:fee1:60c7:cb8a: icmp_seq=4 ttl=64 time=1314 ms
48 bytes from bbbb::ae8d:fee1:60c7:cb8a: icmp_seq=5 ttl=64 time=352 ms

--- bbbb:0:0:0:ae8d:fee1:60c7:cb8a ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 20011ms
rtt min/avg/max/mdev = 175.737/664.487/1313.505/494.614 ms

```

```
14:52:37 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
14:54:19 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:e052:d7d7
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
14:54:43 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:e052:d7d7
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
14:55:03 INFO 7d12 [ICMPv6ECHO] received an echo request (length: 48)
14:55:07 INFO 7d12 [ICMPv6ECHO] received an echo request (length: 48)
14:55:13 INFO 7d12 [ICMPv6ECHO] received an echo request (length: 48)
14:55:17 INFO 7d12 [ICMPv6ECHO] received an echo request (length: 48)
14:55:21 INFO 7d12 [ICMPv6ECHO] received an echo request (length: 48)
14:55:36 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:e052:d7d7
        - parents:
           bbbb:0:0:0:ae8d:fee1:6019:63ee
        - children:
14:55:41 INFO 5de6 [ICMPv6ECHO] received an echo request (length: 48)
14:55:47 INFO 5de6 [ICMPv6ECHO] received an echo request (length: 48)
14:55:51 INFO 5de6 [ICMPv6ECHO] received an echo request (length: 48)
14:55:57 INFO 5de6 [ICMPv6ECHO] received an echo request (length: 48)
14:56:01 INFO 5de6 [ICMPv6ECHO] received an echo request (length: 48)

```
</details>

<details><summary>3 nrf52840based boards</summary>

Note that CRC errors are explained in the doc and do to serial issues.

```
ping6 -s 40 -c 5 -i 5  bbbb:0:0:0:aa04:8169:befa:4338
PING bbbb:0:0:0:aa04:8169:befa:4338(bbbb::aa04:8169:befa:4338) 40 data bytes
48 bytes from bbbb::aa04:8169:befa:4338: icmp_seq=2 ttl=64 time=2081 ms
48 bytes from bbbb::aa04:8169:befa:4338: icmp_seq=3 ttl=64 time=3135 ms
48 bytes from bbbb::aa04:8169:befa:4338: icmp_seq=5 ttl=64 time=3127 ms

--- bbbb:0:0:0:aa04:8169:befa:4338 ping statistics ---
5 packets transmitted, 3 received, 40% packet loss, time 20210ms
rtt min/avg/max/mdev = 2081.356/2781.201/3134.968/494.875 ms

```

```
14:58:46 SUCCESS 6707 [CJOIN] node joined
14:58:58 INFO 6707 [SIXTOP] sending a 6top request
14:59:00 SUCCESS 6707 [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
14:59:24 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
14:59:59 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:00:23 WARNING 6707 [IEEE802154E] large timeCorr.: -8 ticks (code loc. 0)
15:00:35 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:00:49 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:01:13 WARNING 6707 [IEEE802154E] large timeCorr.: -7 ticks (code loc. 0)
15:01:36 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 1)
15:01:58 WARNING 6707 [IEEE802154E] large timeCorr.: -9 ticks (code loc. 1)
15:02:02 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:02:56 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:03:18 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 1)
15:04:23 INFO received RPL DAO from bbbb:0:0:0:aa04:8169:befa:4338
        - parents:
           bbbb:0:0:0:6ed6:5bb6:4b52:c52f
        - children:
15:04:45 WARNING 6707 [IEEE802154E] large timeCorr.: -9 ticks (code loc. 1)
15:04:45 INFO received RPL DAO from bbbb:0:0:0:aa04:8169:befa:4338
        - parents:
           bbbb:0:0:0:6ed6:5bb6:4b52:c52f
        - children:
15:04:58 ERROR 25f0 [OPENSERIAL] wrong CRC in input Buffer
15:05:01 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 1)
15:05:01 INFO received RPL DAO from bbbb:0:0:0:aa04:8169:befa:4338
        - parents:
           bbbb:0:0:0:6ed6:5bb6:4b52:c52f
        - children:
15:05:04 INFO 6707 [ICMPv6ECHO] received an echo request (length: 48)
15:05:10 INFO 6707 [ICMPv6ECHO] received an echo request (length: 48)
15:05:13 ERROR 25f0 [OPENSERIAL] wrong CRC in input Buffer
15:05:17 WARNING 6707 [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:05:20 INFO 6707 [ICMPv6ECHO] received an echo request (length: 48)

```
</details>

<details><summary>2 iotlab-m3 (does not use the radio hal ATM)</summary>

Note that CRC errors are explained in the doc and do to serial issues.

```
15:04:46 SUCCESS Setting mote 594f as root
15:04:46 INFO starting RPC server
15:04:46 ERROR 594f [IEEE802154E] wrong state 1 in startSlot, at slotOffset 1
15:04:47 INFO registering DAGroot ba-f0-3b-5d-23-89-d9-4f
15:05:14 SUCCESS 1f2a [IEEE802154E] synchronized at slotOffset 0
15:05:38 INFO 1f2a [CJOIN] sending CJOIN request
15:05:39 VERBOSE New node: ea5c:d4f:eba2:1f2a. Creating new OSCORE context in /home/francisco/workspace/RIOT/bin/oscore_context_ea5c0d4feba21f2a.json.
15:05:39 VERBOSE received JRC join request
15:05:39 ERROR 594f [OPENSERIAL] wrong CRC in input Buffer
15:05:51 WARNING 1f2a [IEEE802154E] large timeCorr.: 6 ticks (code loc. 0)
15:06:37 INFO 1f2a [CJOIN] sending CJOIN request
O*:06:38 ERROR Type-error in conversion of \
15:06:38 VERBOSE received JRC join request
15:06:39 SUCCESS 1f2a [CJOIN] node joined
15:07:10 INFO 1f2a [SIXTOP] sending a 6top request
15:07:11 WARNING 1f2a [IEEE802154E] large timeCorr.: -6 ticks (code loc. 0)
15:07:11 SUCCESS 1f2a [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
15:07:53 WARNING 1f2a [IEEE802154E] large timeCorr.: 10 ticks (code loc. 1)
15:08:13 INFO received RPL DAO from bbbb:0:0:0:ea5c:d4f:eba2:1f2a
        - parents:
           bbbb:0:0:0:baf0:3b5d:2389:d94f
        - children:
15:08:26 INFO 1f2a [ICMPv6ECHO] received an echo request (length: 48)
15:08:30 INFO 1f2a [ICMPv6ECHO] received an echo request (length: 48)
15:08:36 INFO 1f2a [ICMPv6ECHO] received an echo request (length: 48)
15:08:40 INFO 1f2a [ICMPv6ECHO] received an echo request (length: 48)
15:08:46 INFO 1f2a [ICMPv6ECHO] received an echo request (length: 48)

```

```
ping6 -s 40 -c 5 -i 5 bbbb:0:0:0:ea5c:d4f:eba2:1f2a
PING bbbb:0:0:0:ea5c:d4f:eba2:1f2a(bbbb::ea5c:d4f:eba2:1f2a) 40 data bytes
48 bytes from bbbb::ea5c:d4f:eba2:1f2a: icmp_seq=1 ttl=64 time=3224 ms
48 bytes from bbbb::ea5c:d4f:eba2:1f2a: icmp_seq=2 ttl=64 time=2252 ms
48 bytes from bbbb::ea5c:d4f:eba2:1f2a: icmp_seq=3 ttl=64 time=3314 ms
48 bytes from bbbb::ea5c:d4f:eba2:1f2a: icmp_seq=4 ttl=64 time=2348 ms
48 bytes from bbbb::ea5c:d4f:eba2:1f2a: icmp_seq=5 ttl=64 time=3410 ms

--- bbbb:0:0:0:ea5c:d4f:eba2:1f2a ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 20006ms
rtt min/avg/max/mdev = 2252.446/2909.741/3410.362/502.221 ms
```
</details>

File: cpu/cc2538/include/cpu_conf.h [hunk 11]
```diff
@@ -43,6 +43,19 @@ extern "C" {
...
+ * @name    OpenWSN timing constants
+ *
+ * @{
+ */
+/* Taken from openwsn-fw */
+#define PORT_maxTxDataPrepare   (460/PORT_US_PER_TICK)
+#define PORT_maxRxAckPrepare    (300/PORT_US_PER_TICK)
+#define PORT_maxRxDataPrepare   (300/PORT_US_PER_TICK)
+#define PORT_maxTxAckPrepare    (460/PORT_US_PER_TICK)
+#define PORT_delayTx            (400/PORT_US_PER_TICK)
```
Comment (bergzand) [inline cpu/cc2538/include/cpu_conf.h:56]: Does it make sense to rework these to RIOT-specific defines and keep the openwsn-specific defines in the package?

So here an:
```
#define RADIO_MAX_TX_DATA_PREPARE   (460/PORT_US_PER_TICK)
#define RADIO_MAX_RX_ACK_PREPARE    (300/PORT_US_PER_TICK)
#define RADIO_MAX_RX_DATA_PREPARE   (300/PORT_US_PER_TICK)
#define RADIO_MAX_TX_ACK_PREPARE    (460/PORT_US_PER_TICK)
#define RADIO_DELAY_TX                     (400/PORT_US_PER_TICK)
```

And in the OpenWSN define the OpenWSN-specific defines to these values. This way we can easily reuse these values for other MAC protocols.

Comment (fjmolinas) [reply cpu/cc2538/include/cpu_conf.h:56]: Not really, maybe the ls two `RADIO_DELAY_TX` and `RADIO_DELAY_RX`, the other names are specific to OpenWSN state machine, its a mix of hardware and code introduced delays https://openwsn.atlassian.net/wiki/spaces/OW/pages/688255/Timing+Constants.

Comment (jia200x) [reply cpu/cc2538/include/cpu_conf.h:56]: Are this CPU configurations or board configurations?

**Comment by jia200x** [comment]
the test results look good to me. Please squash :)

File: cpu/cc2538/include/cpu_conf.h (hunk 11 above)
Comment (bergzand) [reply cpu/cc2538/include/cpu_conf.h:56]: They should be CPU configuration in the nrf52840 case right? I think the same holds for the cc2538, or any CPU with integrated radio.

Comment (fjmolinas) [reply cpu/cc2538/include/cpu_conf.h:56]: > Are this CPU configurations or board configurations?

CPU since its an integrated radio, for the `iotlab-m3` and `samr21-xpro` they are BOARD configurations.

**Review by jia200x — APPROVED** [review]
ACK. I trust the test results posted in https://github.com/RIOT-OS/RIOT/pull/15145#issuecomment-715330303 and the radio HAL integration looks solid.

File: cpu/cc2538/include/cpu_conf.h (hunk 11 above)
Comment (jia200x) [reply cpu/cc2538/include/cpu_conf.h:56]: at some point we would like to have a way to ask the hardware about its timings. Maybe that should be part of the Radio HAL, but we are quite far from there (:

**Review by jia200x — APPROVED** [review]
Re-ACK

**Comment by fjmolinas** [comment]
Thanks a lot for the review @jia200x and the offline help! Much appreciated! And great work with the radio-hal, was pretty easy to fit OpenWSN on top of it!
